### ISSUE TITLE
Issues with Managed Identity Authentication via Azure DevOps - DB connection failed after 30s timeout

### ISSUE THREAD

**Issue by mischmuc** [betr-io/terraform-provider-mssql#91]
Previously, I was using the provider with an Azure DevOps Service Account connected via Service Principal, and it was working fine. Now, I replaced it with a Service Account using Managed Identity with Federated Credentials and updated the code to use azuread_managed_identity_auth with user_id as described in the provider documentation. Unfortunately, it is not working.

I initially thought it was a firewall issue, but my DevOps Agent IP is set in the firewall rules, and the managed identity is assigned as Entra admin.

I still receive the error: "Unable to create user [db-name].[mi-user]: DB connection failed after 30s timeout."

I am using the following ID for user_id: /subscriptions/.../resourcegroups/rg-name/providers/Microsoft.ManagedIdentity/userAssignedIdentities/mi-name.

If I switch back to the Service Principal, it works again. All my other resources are also working fine with the Managed Identity.
Any idea what the problem could be?

```
resource "mssql_user" "mi_db_user" {
  for_each = local.managed_identity

  server {
      host = data.azurerm_mssql_server.sql_server.fully_qualified_domain_name
      azuread_managed_identity_auth {
         user_id = local.mi_admin_id_string
      }
    }

  database       = data.azurerm_mssql_database.sql_database.name
  username      = each.value.sql_username
  roles              = ["db_datareader", "db_datawriter", "db_executor"]
  default_schema = "dbo"
  
  depends_on = [data.azurerm_mssql_server.sql_server, azuread_group_member.sqlserver_member]
}
```

**Comment by korneel76** [comment]
I have got the same issue. Whatever I try, I always get the same error:
 Error: unable to create user [xxx].[xxx]: db connection failed after 30s timeout.

@mischmuc, you say it works for you when you use the provider with an Azure DevOps Service Account connected via Service Principal. When I do that I still get the timeout.
What does your resource look like in your old situation? The only thing you change is "azuread_managed_identity_auth" ?

**Comment by mischmuc** [comment]
@korneel76 Yes, with  via Service Principal it is working for me, you have to use 
````
server {
      host = "your_sql_server_fully_qualified_domain_name"
         azure_login {
            tenant_id = "your_tenant_id"
            client_id = "your_sp_client_id"
            client_secret = "your_sp_client_secret"
         }
}
````

In addition at least in my setup I'm using the same Service Principal as EntraID Admin on the Server and for the permissions to AD (EntraId) like described here: https://learn.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal?view=azuresql

My Server is running behind a Vnet / Firewall so I also have to open the firewall for the current DevOps Agent IP.

Hope this helps

**Comment by korneel76** [comment]
@mischmuc Thanks. I tried the azure_login already, but I'm not sure about the permissions. So I will try again. ;-)

**Comment by Roymprog** [comment]
@mischmuc, I'm experiencing exactly the same when switching to a user managed identity. Did you manage to find a workaround?

**Comment by mischmuc** [comment]
Unfortunately not but I don't have time to test it again. I switched meanwhile to a dacpac approach outside terraform

**Comment by nikolaof** [comment]
I'm dealing with the same issue here. More specifically my setup was the following:

```
resource "mssql_user" "aad_user" {
  for_each = var.databases

  server {
    host = "${var.mssql_server_name}.database.windows.net"

    azure_login {
      # Leave tenant_id, client_id, and client_secret arguments out.
      # The MSSQL provider will automatically read
      # MSSQL_TENANT_ID, MSSQL_CLIENT_ID, and MSSQL_CLIENT_SECRET
      # from the environment.
    }
  }

  database  = azurerm_mssql_database.db_instance[each.key].name
  username  = azurerm_user_assigned_identity.service_identity[each.key].name
  object_id = azurerm_user_assigned_identity.service_identity[each.key].client_id

  roles = ["db_owner"]
}
```

And I have exported the corresponding env variables with a service principal info.

The first time that I applied the code it worked and applied everything correctly. But then I made some changes and tried to re-apply and stuck into the same error:

```
unable to read user [x].[x]: db connection failed after 30s timeout
```

Here are also some facts:

The first time that I had to run it I forgot to whitelist my IP and I got a corresponding error that my IP is blocked. I whitelisted and worked. Now after this first time I removed my IP from the firewall to see if this will trigger the same error BUT NOT. Provider still spits out  the 

```
unable to read user [x].[x]: db connection failed after 30s timeout
```

Even if I tried to hardcode the credentials in the `azure_login` block it didn't work.

Some logs from Terraform shows the following:

```
025-12-03T20:32:12.806Z [INFO]  provider.terraform-provider-mssql_v0.3.1: 2025/12/03 20:32:12 failed to connect to database: parameter 'clientID' cannot be empty: timestamp=2025-12-03T20:32:12.806Z
2025-12-03T20:32:14.409Z [ERROR] provider.terraform-provider-mssql_v0.3.1: Response contains error diagnostic: diagnostic_detail="" diagnostic_summary="unable to read user [x].[x]: db connection failed after 30s timeout" tf_req_id=18dc45d6-4e2f-0140-6b26-3175ae2d01c8 @module=sdk.proto diagnostic_severity=ERROR tf_proto_version=5.4 tf_provider_addr=provider tf_resource_type=mssql_user tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/diag/diagnostics.go:62 timestamp=2025-12-03T20:32:14.409Z
```

**Comment by audunsolemdal** [comment]
After trying the same (although via Github Actions federated credentials). From what I can read from the code this provider currently doesn't support the scenario.

When using federated credentials via e.g. Github Actions or ADO pipelines, you do not use ManagedIdentityCredential, but WorkloadIdentityCredential (I was also mislead by the naming when I first encountered it)

From what I can see this provider depends on [this driver](https://github.com/denisenkom/go-mssqldb#azure-active-directory-authentication) which relies on [azidentity v0.11.0,](https://github.com/denisenkom/go-mssqldb/blob/master/go.mod#L7) which is older than when support for [WorkloadIdentityCredential](https://github.com/Azure/azure-sdk-for-go/pull/19503)   was released in version 1.3.0

**Comment by audunsolemdal** [comment]
> After trying the same (although via Github Actions federated credentials). From what I can read from the code this provider currently doesn't support the scenario.
> 

I managed to get it working from Github Actions with this provider though if you are still interested.

https://registry.terraform.io/providers/muecahit94/mssql/latest/

```hcl
provider "mssql" {
  hostname = "xxx"
  port     = 1433
  azure_auth {}
}

resource "mssql_azuread_user" "this" {

  database_name  = "${azurerm_resource_group.rg.name}db"
  name           = "name-of-entragroup"
  object_id      = "xxx" # fails with "Assign an Azure AD identity to your server and add Directory Reader permission to your identity\" (https://aka.ms/sqlaadsetup)' if not set
  default_schema = "dbo"

  roles = [
    "db_datareader",
  ]
}

resource "mssql_azuread_user" "appservice_user" {

  database_name = "${azurerm_resource_group.rg.name}db"
  name          = "name-of-appservice"
  object_id      = "xxx" # fails with "Assign an Azure AD identity to your server and add Directory Reader permission to your identity\" (https://aka.ms/sqlaadsetup)' if not set
  roles         = ["db_datareader"]
}
```

you need to have the env vars AZURE_CLIENT_ID and AZURE_TENANT_ID set + the env var AZURE_FEDERATED_TOKEN_FILE needs to point towards a file containing your Pipeline's/workflow's token
